xrtl/testing/render_test.cc

```diff
@@ -0,0 +1,98 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "xrtl/testing/render_test.h"
+
+// TODO(scotttodd): Move this to some build magic and/or a .cc file
+#define STBI_NO_LINEAR  // no loadf
+#define STBI_NO_HDR     // no hdr conversion
+#define STBI_ONLY_PNG   // only .png support
+#define STB_IMAGE_IMPLEMENTATION
+#include "stb_image.h"
+
+#include "xrtl/base/logging.h"
+
+namespace xrtl {
+namespace testing {
+
+bool RenderTest::CompareImages(const std::string& golden_image_path,
+                               const std::string& test_image_path) {
+  // TODO(scotttodd): Wrap this in an object.
+  int test_image_width = 0;
+  int test_image_height = 0;
+  int test_image_channels = 0;
+  uint8_t* test_image_data =
+      stbi_load(test_image_path.c_str(), &test_image_width, &test_image_height,
+                &test_image_channels, /*desired_channels=*/3);
+  if (!test_image_data) {
+    LOG(ERROR) << "Could not load the test image";
+    return false;
+  }
+
+  bool result = CompareImages(golden_image_path, test_image_data,
+                              test_image_width, test_image_height);
+
+  stbi_image_free(test_image_data);
+  return result;
+}
+
+bool RenderTest::CompareImages(const std::string& golden_image_path,
+                               uint8_t* test_image_data, int test_image_width,
+                               int test_image_height) {
+  // TODO(scotttodd): Output diff image and test image to approve if different.
+  // TODO(scotttodd): Diff percentage.
+
+  // TODO(scotttodd): Wrap this in an object.
+  int golden_image_width = 0;
+  int golden_image_height = 0;
+  int golden_image_channels = 0;
+  uint8_t* golden_image_data = stbi_load(
+      golden_image_path.c_str(), &golden_image_width, &golden_image_height,
+      &golden_image_channels, /*desired_channels=*/3);
+  if (!golden_image_data) {
+    LOG(ERROR) << "Could not load the golden image";
+    return false;
+  }
+
+  if (test_image_width != golden_image_width) {
+    LOG(ERROR) << "Image widths don't match. Expected " << golden_image_width
+               << "px but test was " << test_image_width << "px";
+    stbi_image_free(golden_image_data);
```

Comment:
you could wrap the data in a unique_ptr to avoid the need for the manual memory tracking. for example:

```
std::unique_ptr<uint8_t, void(*)(uint8_t*)> golden_image_data{stbi_load(...), stbi_image_free};
```

Could make this a helper function so you can just auto golden_image_data = ...;